### PR TITLE
Change addresses to net.Addr

### PR DESCRIPTION
**Pull request by emersion**
The motivation is to allow Header to hold Unix socket addresses.

Instead of using an IP and a port, use net.Addr, which allows to
represent any network address.

The Header.{Local,Remote}Addr methods are removed since users can now
just access Header.{Source,Destination}Addr.

New helpers (TCPAddrs, UDPAddrs, UnixAddrs, IPs, Ports) allows users to
easily extract data from the generic net.Addr fields.

Closes: https://github.com/pires/go-proxyproto/issues/34
Closes: https://github.com/pires/go-proxyproto/issues/43

* * *

Tests haven't been updated yet, but I'd like to gather some feedback because this is a quite radical change. What do you think?

### CODE REVIEW

File: v2.go [hunk 1]
```diff
@@ -107,19 +108,15 @@ func parseVersion2(reader *bufio.Reader) (header *Header, err error) {
...
 			return nil, ErrInvalidAddress
 		}
-		header.SourceAddress = addr.Src[:]
-		header.DestinationAddress = addr.Dst[:]
-		header.SourcePort = addr.SrcPort
-		header.DestinationPort = addr.DstPort
+		header.SourceAddr = newIPAddr(header.TransportProtocol, addr.Src[:], addr.SrcPort)
+		header.DestinationAddr = newIPAddr(header.TransportProtocol, addr.Dst[:], addr.DstPort)
 	}
 	// TODO fully support Unix addresses
```
Comment (pires) [inline v2.go:121]: Uncomment this block?

File: v2.go [hunk 2]
```diff
@@ -157,38 +154,43 @@ func (header *Header) formatVersion2() ([]byte, error) {
...
 	buf.Write(addrDst)
 
-	portSrcBytes := func() []byte {
-		a := make([]byte, 2)
-		binary.BigEndian.PutUint16(a, header.SourcePort)
-		return a
-	}()
-	buf.Write(portSrcBytes)
+	if sourcePort, destPort, ok := header.Ports(); ok {
+		portBytes := make([]byte, 2)
```
Comment (pires) [inline v2.go:185]: Shouldn't it be 4 instead?

**Review by pires — COMMENTED** [review]
It's looking good to me so far.

File: v2.go (hunk 2 above)
Comment (emersion) [reply v2.go:185]: The part is an `uint16`, so 2 bytes, unless I'm missing something? `portBytes` is overwritten for each port (source, then dest).

I suppose we could instead put the source port in `portBytes[0:2]` and the destination port in `portBytes[2:4]`, then write the whole thing.

Comment (pires) [reply v2.go:185]: Nevermind, I had misread the code and it's fine to overwrite as there's a write to the buffer before that.

**Comment by emersion** [comment]
All done! This is ready for another review.

**Review by pires — CHANGES_REQUESTED** [review]
Overall LGTM but we're missing changes to `README.md` to reflect the breaking changes.

Also, I'll need a bit more time to review this properly.

Thank you very much for the work you've put into this 🤩

**Comment by emersion** [comment]
>we're missing changes to README.md to reflect the breaking changes

Fixed

>Also, I'll need a bit more time to review this properly.

Sure, take your time :)

>Thank you very much for the work you've put into this

Thanks for maintaining this library!

**Review by pires — APPROVED** [review]
🎉

**Pull request merged**